src/coreclr/jit/gentree.cpp

```diff
@@ -25399,8 +25386,620 @@ GenTree* Compiler::gtNewSimdRoundNode(var_types type, GenTree* op1, CorInfoType
+                                                simdSize);
+            }
+            else
+            {
+                // if we have a constant, keep it constant
+                GenTree* op1Dup1 = fgMakeMultiUse(&op1);
+                swap             = op1Dup1;
+
+                simd_t*  cnsPtr = &op1Dup1->AsVecCon()->gtSimdVal;
+                std::swap(cnsPtr->u64[0], cnsPtr->u64[2]);
+                std::swap(cnsPtr->u64[1], cnsPtr->u64[3]);
+            }
+
+            // shuffle with both the normal and swapped values
+            // Vector256<byte> shuf1 = Avx2.Shuffle(vector, indices);
+            // Vector256<byte> shuf2 = Avx2.Shuffle(swap, indices);
+            GenTree* shuf1 = gtNewSimdHWIntrinsicNode(type, op1, op2, NI_AVX2_Shuffle, simdBaseJitType, simdSize);
+            GenTree* shuf2 = gtNewSimdHWIntrinsicNode(type, swap, op2Dup1, NI_AVX2_Shuffle, simdBaseJitType, simdSize);
+
+            // get the indices, and xor the cross-lane bit on the high 128-bit lane part of indices.
+            // V256<byte> indicesXord = indices ^ V256.Create(V128.Create((byte)0), V128.Create((byte)0x10)));
+            simd_t xorCns                  = {};
+            xorCns.u64[0]                  = 0;
+            xorCns.u64[1]                  = 0;
+            xorCns.u64[2]                  = 0x1010101010101010;
+            xorCns.u64[3]                  = 0x1010101010101010;
+            cnsNode                        = gtNewVconNode(type);
+            cnsNode->AsVecCon()->gtSimdVal = xorCns;
+
+            GenTree* indicesXord = gtNewSimdBinOpNode(GT_XOR, type, op2Dup2, cnsNode, simdBaseJitType, simdSize);
+
+            // compare our modified indices to 0x0F (highest value not swapping lane), we get 0xFF when we are swapping
+            // lane and 0x00 otherwise. we will also get "swapping lane" also when index is more than 32 (but no high bit),
+            // but this is normalised later for Shuffle, and acceptable for ShuffleNative.
+            // V256<byte> selection = Avx2.CompareGreaterThan(indicesXord.AsSByte(), V256.Create((sbyte)0x0F)).AsByte();
+            simd_t comparandCnd            = {};
+            comparandCnd.u64[0]            = 0x0F0F0F0F0F0F0F0F;
+            comparandCnd.u64[1]            = 0x0F0F0F0F0F0F0F0F;
+            comparandCnd.u64[2]            = 0x0F0F0F0F0F0F0F0F;
+            comparandCnd.u64[3]            = 0x0F0F0F0F0F0F0F0F;
+            cnsNode                        = gtNewVconNode(type);
+            cnsNode->AsVecCon()->gtSimdVal = comparandCnd;
+            GenTree* selection = gtNewSimdCmpOpNode(GT_GT, type, indicesXord, cnsNode, CORINFO_TYPE_BYTE, simdSize);
+
+            // blend our two shuffles based on whether each element swaps lanes or not
+            // return Avx2.BlendVariable(shuf1, shuf2, selection);
+            retNode = gtNewSimdHWIntrinsicNode(type, shuf1, shuf2, selection, NI_AVX2_BlendVariable, simdBaseJitType,
```

Comment:
@tannergooding for when you review: when I was originally testing this, this code had better performance when it was ordered slightly differently to here (due to better pipelining). Ideally we start `swap` before `shuf1` and `indicesXord` before `shuf1` iirc. Please advise if this is possible and how to do it. Presumably it's something with `gtWrapWithSideEffects`(?), but I couldn't quite figure it out.